### PR TITLE
fix bug 1182210 - Store bitly URLs on model

### PR DESCRIPTION
**Pull request by robhudson**

### CODE REVIEW

File: kuma/core/tests/test_utils.py [hunk 1]
```diff
@@ -21,3 +25,26 @@ def test_empty_string(self):
     def test_wrong_type(self):
         eq_(0, smart_int(None))
         eq_(10, smart_int([], 10))
+
+
+@override_settings(USE_SHORT_URLS=True)
+class BitlyTestCase(KumaTestCase):
```
Comment (jezdez) [inline kuma/core/tests/test_utils.py]: The `KumaTestCase` isn't needed here and we can just use Django's standard test class

File: kuma/wiki/migrations/0008_auto_20150921_1154.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
...), (b'ha', 'Hausa'), (b'he', '\u05e2\u05d1\u05e8\u05d9\u05ea'), (b'hi-IN', '\u0939\u093f\u0928\u094d\u0926\u0940 (\u092d\u093e\u0930\u0924)'), (b'hr', 'Hrvatski'), (b'hu', 'magyar'), (b'id', 'Bahasa Indonesia'), (b'ig', 'Igbo'), (b'it', 'Italiano'), (b'ja', '\u65e5\u672c\u8a9e'), (b'ka', '\u10e5\u10d0\u10e0\u10d7\u10e3\u10da\u10d8'), (b'ko', '\ud55c\uad6d\uc5b4'), (b'ln', 'Ling\xe1la'), (b'mg', 'Malagasy'), (b'ml', '\u0d2e\u0d32\u0d2f\u0d3e\u0d33\u0d02'), (b'ms', 'Melayu'), (b'my', '\u1019\u103c\u1014\u103a\u1019\u102c\u1018\u102c\u101e\u102c'), (b'nl', 'Nederlands'), (b'pl', 'Polski'), (b'pt-BR', 'Portugu\xeas (do\xa0Brasil)'), (b'pt-PT', 'Portugu\xeas (Europeu)'), (b'ro', 'rom\xe2n\u0103'), (b'ru', '\u0420\u0443\u0441\u0441\u043a\u0438\u0439'), (b'son', 'So\u014bay'), (b'sq', 'Shqip'), (b'sv-SE', 'Svenska'), (b'sw', 'Kiswahili'), (b'ta', '\u0ba4\u0bae\u0bbf\u0bb4\u0bcd'), (b'th', '\u0e44\u0e17\u0e22'), (b'tl', 'Tagalog'), (b'tn', 'Setswana'), (b'tr', 'T\xfcrk\xe7e'), (b'vi', 'Ti\u1ebfng Vi\u1ec7t'), (b'wo', 'Wolof'), (b'xh', 'isiXhosa'), (b'yo', 'Yor\xf9b\xe1'), (b'zh-CN', '\u4e2d\u6587 (\u7b80\u4f53)'), (b'zh-TW', '\u6b63\u9ad4\u4e2d\u6587 (\u7e41\u9ad4)'), (b'zu', 'isiZulu')]),
```
Comment (jezdez) [inline kuma/wiki/migrations/0008_auto_20150921_1154.py]: I think it would be nicer to have thise locale related change in a separate migration

File: kuma/wiki/models.py [hunk 3]
```diff
@@ -1467,6 +1473,32 @@ def contributors(self):
     def zone_stack(self):
         return DocumentZoneStackJob().get(self.pk)
 
+    def set_share_url(self, save=True):
+        share_url = bitly_shorten('/'.join([settings.SITE_URL,
```
Comment (jezdez) [inline kuma/wiki/models.py]: Please use `urlparse.urljoin` here

File: kuma/wiki/models.py [hunk 4]
```diff
@@ -1467,6 +1473,32 @@ def contributors(self):
...
+        if save and share_url:
+            self.share_url = share_url
+            if self.pk:
+                self.save(update_fields=['share_url'])
+            else:
+                self.save()
+
+    def get_share_url(self):
+        """
+        Returns the `share_url` or generates it if None.
```
Comment (jezdez) [inline kuma/wiki/models.py]: Inline code is two backticks in rST:

    ``share_url``

File: kuma/wiki/models.py [hunk 5]
```diff
@@ -1467,6 +1473,32 @@ def contributors(self):
...
+
+        If not `settings.USE_SHORT_URLS`, we return the full URL.
+
+        """
+        if settings.USE_SHORT_URLS:
+            if self.share_url:
+                return self.share_url
+
+            self.set_share_url()
```
Comment (jezdez) [inline kuma/wiki/models.py]: let's add the else clause here for clarity

File: kuma/wiki/models.py [hunk 6]
```diff
@@ -1467,6 +1473,32 @@ def contributors(self):
...
+
+        else:
+            return '/'.join([settings.SITE_URL, self.get_absolute_url()])
```
Comment (jezdez) [inline kuma/wiki/models.py]: urlparse.urljoin again please

File: kuma/wiki/templates/wiki/includes/document_macros.html [hunk 7]
```diff
@@ -85,26 +85,22 @@
...
 
-    {% set short_share_url_twitter = share_url_twitter | bitly_shorten %}
-    {% set short_share_url_facebook = share_url_facebook | bitly_shorten %}
-    {% set short_share_url_gplus = share_url_gplus | bitly_shorten %}
-
-    {% set twitter_url = 'http://twitter.com/share'|urlparams(url=short_share_url_twitter, text=share_text) %}
-    {% set facebook_url = 'http://facebook.com/sharer.php?s=100&p[url]=' + short_share_url_facebook %}
-    {% set gplus_url = 'https://plus.google.com/share'|urlparams(url=short_share_url_gplus) %}
+    {% set twitter_url = 'http://twitter.com/share'|urlparams(url=share_url_twitter, text=share_text) %}
+    {% set facebook_url = 'http://facebook.com/sharer.php?s=100&p[url]=' + share_url_facebook %}
```
Comment (jezdez) [inline kuma/wiki/templates/wiki/includes/document_macros.html]: I was wondering if we should replace `&` with `&amp;`?

File: kuma/core/utils.py [hunk 8]
```diff
@@ -321,3 +326,15 @@ def get_unique(content_type, object_pk, name=None, request=None,
...
+
+def bitly_shorten(url):
+    if not settings.USE_SHORT_URLS:
+        return None
+
+    try:
+        return bitly.shorten(url)['url']
+    except (bitly_api.BitlyError, KeyError) as e:
+        # If bit.ly service fails or the API key isn't configured.
+        log.error('Bitly URL shortening failed: %s' % e)
```
Comment (jezdez) [inline kuma/core/utils.py]: I think we should use `log.exception` here (this was a bug in the previous code as well). Without it we don't get the traceback object passed to the logger.

File: kuma/demos/helpers.py [hunk 9]
```diff
@@ -345,3 +347,24 @@ def _contextual_locale(context):
     if not localedata.exists(locale):
         locale = settings.LANGUAGE_CODE
     return locale
+
+
+bitly = bitly_api.Connection(login=getattr(settings, 'BITLY_USERNAME', ''),
```
Comment (jezdez) [inline kuma/demos/helpers.py]: Couldn't we reuse the core bitly client connection here?

File: kuma/wiki/models.py [hunk 10]
```diff
@@ -920,6 +923,9 @@ def save(self, *args, **kwargs):
             # If this is a translation without a topic parent, try to get one.
             self.acquire_translated_topic_parent()
 
+        # Set the shortened URL (e.g. bit.ly).
+        self.set_share_url(save=False)
```
Comment (jezdez) [inline kuma/wiki/models.py]: I thought we'd run this as a Celery task to be honest to not make document saving slower and potentially blocking on a call to Bitly's API.

**Comment by jezdez** [comment]
@robhudson looking good for the most part, comments and questions inline :)

Also, I'd like to suggest to push the branch to the main repo, as it has positive side effects on shared responsibility of the branch and allows easy collaboration in the code. It prevents having to close and reopen PRs when a handover is done over the course of the life of a branch.

**Comment by darkwing** [comment]
w00t!  YES!

File: kuma/wiki/models.py (hunk 4 above)
Comment (robhudson) [inline kuma/wiki/models.py]: I never understood when to use a single backtick and when to use a double backticks. It's not so much inline code as a reference to a variable.

File: kuma/wiki/templates/wiki/includes/document_macros.html (hunk 7 above)
Comment (robhudson) [inline kuma/wiki/templates/wiki/includes/document_macros.html]: I don't think so, I think that would break the URL.

File: kuma/core/utils.py (hunk 8 above)
Comment (robhudson) [inline kuma/core/utils.py]: :+1:

File: kuma/wiki/models.py (hunk 10 above)
Comment (robhudson) [inline kuma/wiki/models.py]: We have 2 places where we're calling this. On save (here) and on document display if it's not set already. Would we want tasks for both?

On save it seems like we aren't too concerned with speed but we're also not concerned with a delay in getting the bitly URL.

On display we don't want to block but also what do we do assuming the task doesn't get the bitly URL in time... I could trigger the task and return the full URL. The next time it is loaded the document would have the bitly URL.

File: kuma/wiki/models.py (hunk 4 above)
Comment (jezdez) [inline kuma/wiki/models.py]: The best way I think as it would allow autodoc sections is this style: http://sphinx-doc.org/domains.html#info-field-lists

File: kuma/wiki/models.py (hunk 10 above)
Comment (jezdez) [inline kuma/wiki/models.py]: Yeah, that would work for me, as it's the same pattern we use already for the tidied content (see `get_tidied_content`)

File: kuma/wiki/templates/wiki/includes/document_macros.html (hunk 7 above)
Comment (jezdez) [inline kuma/wiki/templates/wiki/includes/document_macros.html]: Ah, oki

**Comment by robhudson** [comment]
Updated based on feedback.

What's missing is adding a management command to process all (or most?) documents to generate the bitly URLs as celery tasks.

File: kuma/wiki/models.py [hunk 11]
```diff
@@ -925,6 +929,10 @@ def save(self, *args, **kwargs):
         # Delete any cached last-modified timestamp.
         self.fill_last_modified_cache()
 
+        # Set the shortened URL (e.g. bit.ly).
+        # TODO: Should we do this on a subset of documents? E.g. not redirects?
+        tasks.set_share_url.delay(self.pk)
```
Comment (groovecoder) [inline kuma/wiki/models.py]: We shouldn't do this on every call to `Document.save()` right? That means every time we save a new `Revision` and set it to the document `current_revision`, we will generate a brand new share url for the document?

**Comment by groovecoder** [comment]
I suggest we skip a management command to generate share urls for every doc. There's already a TODO noting that redirects do not need share urls; and neither do templates, etc.

So, I suggest we let ordinary site traffic trigger the background tasks to generate share urls as documents are visited. Though we should set a rate limit on the celery task so bitly doesn't shut us down again. :neutral_face:

File: kuma/wiki/models.py (hunk 11 above)
Comment (robhudson) [inline kuma/wiki/models.py]: True. It wouldn't get a new share URL. Bitly is smart enough to return the same one. But you are right, we can make this smarter I believe. It's not as simple as not setting if it one already exists b/c we need to handle the case of tree moving, where docs may have a `share_url` already but the URL is changing so it needs to be updated.

**Comment by robhudson** [comment]
Updated:
* don't call bitly for templates, redirects, and existing documents.
* call bitly for documents during `_move_tree`.
* add celery rate limiting.
* calls to `get_share_url` with no short_url returns full document url as fallback.

File: kuma/wiki/migrations/0008_add_share_url.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='document',
+            name='share_url',
+            field=models.URLField(null=True, blank=True),
+            preserve_default=True,
+        ),
+    ]
```
Comment (jezdez) [inline kuma/wiki/migrations/0008_add_share_url.py]: Oh, no new migration for the locale fields needed?

File: kuma/wiki/models.py [hunk 13]
```diff
@@ -925,6 +931,11 @@ def save(self, *args, **kwargs):
         # Delete any cached last-modified timestamp.
         self.fill_last_modified_cache()
 
+        # Set the shortened URL (e.g. bit.ly). Note: We do this only for new
+        # documents that aren't templates or redirects.
+        if is_new and not self.is_template and not self.is_redirect:
```
Comment (jezdez) [inline kuma/wiki/models.py]: Please put this into the `on_document_save` signal handler in the `wiki/apps.py` and check for the `created` flag passed into the signal handler. That's the canonical place for such things now

File: kuma/wiki/tasks.py [hunk 14]
```diff
@@ -460,3 +461,31 @@ def tidy_revision_content(pk):
             )
         # return the errors so we can look them up in the Celery task result
         return errors
+
+
+@task(rate_limit='10/m')
+def set_share_url(pk):
```
Comment (jezdez) [inline kuma/wiki/tasks.py]: please use a more descriptive name for the task such as `update_document_share_url` or something. the name of the ask will show up in the celery admin and it's then easier to distinguish from other the tasks

File: kuma/wiki/models.py [hunk 15]
```diff
@@ -1467,6 +1484,21 @@ def contributors(self):
     def zone_stack(self):
         return DocumentZoneStackJob().get(self.pk)
 
+    def get_share_url(self):
+        """
+        Returns the `share_url`.
```
Comment (jezdez) [inline kuma/wiki/models.py]: Double backticks here

File: kuma/wiki/tests/test_models.py [hunk 16]
```diff
@@ -333,6 +332,39 @@ def test_redirect_url_works_for_home_path(self):
         d = document(is_redirect=True, html=html)
         eq_(href, d.get_redirect_url())
 
+    @mock.patch('kuma.wiki.tasks.set_share_url')
+    def test_get_share_url_empty(self, _mock):
+        doc = document(title='test', share_url=None)
```
Comment (jezdez) [inline kuma/wiki/tests/test_models.py]: that's odd, the document helper doesn't save the Document by default, so I'm not sure why the mock is being called. also, mind not using a underscored variable names here?

File: kuma/wiki/tests/test_tasks.py [hunk 17]
```diff
@@ -76,3 +81,31 @@ def test_sitemaps_files(self):
...
         for loc in expected_sitemap_locs:
             ok_(loc in index_xml)
+
+
+@override_settings(BITLY_API_KEY='test', BITLY_USERNAME='test')
+class BitlyTestCase(UserTestCase):
+    fixtures = UserTestCase.fixtures + ['wiki/documents.json']
+
+    def setUp(self):
+        self.long_url = 'http://example.com/long-url'
```
Comment (jezdez) [inline kuma/wiki/tests/test_tasks.py]: please call the super setUp

File: settings.py [hunk 18]
```diff
@@ -836,8 +835,8 @@ def JINJA_CONFIG():
 FLAG_REASONS = DEMO_FLAG_REASONS + WIKI_FLAG_REASONS
 
 # bit.ly
-BITLY_API_KEY = "SET ME IN SETTINGS_LOCAL"
-BITLY_USERNAME = "SET ME IN SETTINGS_LOCAL"
+BITLY_API_KEY = None  # Set in settings_local.py.
```
Comment (jezdez) [inline settings.py]: "Set me in settings_local.py"

File: kuma/wiki/tests/test_models.py [hunk 19]
```diff
@@ -113,7 +112,7 @@ def test_document_is_template(self):
 
         assert not d.is_template
 
-        d.slug = 'Template:test'
+        d.slug = '%stest' % TEMPLATE_TITLE_PREFIX
```
Comment (jezdez) [inline kuma/wiki/tests/test_models.py]: nice <3

File: kuma/wiki/models.py (hunk 13 above)
Comment (robhudson) [inline kuma/wiki/models.py]: Yes, that would be much nicer. Thanks.

File: kuma/wiki/migrations/0008_add_share_url.py (hunk 12 above)
Comment (robhudson) [inline kuma/wiki/migrations/0008_add_share_url.py]: they are needed but I was going to make a separate PR for those.

File: kuma/wiki/tests/test_models.py (hunk 16 above)
Comment (robhudson) [inline kuma/wiki/tests/test_models.py]: It's called b/c the `get_share_url()` will trigger it if not already populated.

**Comment by robhudson** [comment]
Updated based on feedback.

**Comment by jezdez** [comment]
r+
